src/mscorlib/shared/System/Span.NonGeneric.cs

```diff
@@ -249,6 +423,128 @@ internal static unsafe bool OrdinalHelper(ReadOnlySpan<char> span, ReadOnlySpan<
             }
         }
 
+        private static unsafe int CompareOrdinalHelper(ReadOnlySpan<char> span, ReadOnlySpan<char> value)
+        {
+            Debug.Assert(value.Length != 0);
+            Debug.Assert(span.Length != 0);
+            int length = Math.Min(span.Length, value.Length);
+
+            fixed (char* ap = &MemoryMarshal.GetReference(span))
+            fixed (char* bp = &MemoryMarshal.GetReference(value))
+            {
+                char* a = ap;
+                char* b = bp;
+
+                if (*a != *b)
+                    goto DiffOffsetStart;
+
+                if (length >= 2)
+                {
+                    if (*(a + 1) != *(b + 1))
+                        goto DiffOffset1;
+
+                    // Since we know that the first two chars are the same,
+                    // we can increment by 2 here and skip 4 bytes.
+                    // This leaves us 8-byte aligned, which results
+                    // on better perf for 64-bit platforms.
+                    length -= 2;
+                    a += 2;
+                    b += 2;
+                }
+
+                // unroll the loop
+#if BIT64
+                while (length >= 12)
+                {
+                    if (*(long*)a != *(long*)b)
+                        goto DiffOffset0;
+                    if (*(long*)(a + 4) != *(long*)(b + 4))
+                        goto DiffOffset4;
+                    if (*(long*)(a + 8) != *(long*)(b + 8))
+                        goto DiffOffset8;
+                    length -= 12;
+                    a += 12;
+                    b += 12;
+                }
+#else // BIT64
+                while (length >= 10)
+                {
+                    if (*(int*)a != *(int*)b) goto DiffOffset0;
+                    if (*(int*)(a + 2) != *(int*)(b + 2)) goto DiffOffset2;
+                    if (*(int*)(a + 4) != *(int*)(b + 4)) goto DiffOffset4;
+                    if (*(int*)(a + 6) != *(int*)(b + 6)) goto DiffOffset6;
+                    if (*(int*)(a + 8) != *(int*)(b + 8)) goto DiffOffset8;
+                    length -= 10; a += 10; b += 10; 
+                }
+#endif // BIT64
+
+                // Fallback loop:
+                // go back to slower code path and do comparison on 4 bytes at a time.
+                // This depends on the fact that the String objects are
+                // always zero terminated and that the terminating zero is not included
+                // in the length. For odd string sizes, the last compare will include
+                // the zero terminator.
+                while (length > 0)
+                {
+                    if (*(int*)a != *(int*)b)
+                        goto DiffNextInt;
```

Comment:
is this correct? what happen if we have length == 1?

Reply:
Again, this is invalid copy&paste from the `String` implementation.